### PR TITLE
Migrate to sbt-typelevel

### PR DESCRIPTION
**Pull request by armanbilge**
Closes https://github.com/typelevel/cats/issues/4143. Fixes https://github.com/typelevel/cats/issues/4062. Fixes https://github.com/typelevel/cats/issues/4149.

Things seem to be working locally. But the battle with CI is just beginning.

Note this PR doesn't deal with the documentation/site at all (they are currently not being checked in CI). I'll prepare that in a follow-up for easier review, although they should probably merge together.

### CODE REVIEW

File: project/plugins.sbt
Comment (armanbilge) [inline project/plugins.sbt:4]: I forgot to mention: in the end, I didn't actually migrate to sbt-typelevel since that would require adding copyright headers to all the files. Instead we use a combination of sbt-typelevel-ci-release and sbt-typelevel-settings. So the build still has a little more boilerplate than it could, but it's okay.

File: core/src/main/scala/cats/Invariant.scala
Comment (armanbilge) [inline core/src/main/scala/cats/Invariant.scala:111]: This is the fix for https://github.com/typelevel/cats/issues/4062. Added `@targetName` as a dummy annotation in Scala 2 so we don't have to split this file (I tried compat `trait`s first and it didn't work).

File: core/src/main/scala/cats/instances/StaticMethods.scala
Comment (armanbilge) [inline core/src/main/scala/cats/instances/StaticMethods.scala:5]: Fixes https://github.com/typelevel/cats/issues/4149.

File: mima.sbt
Comment (armanbilge) [inline mima.sbt:1]: I moved these to their own file so they don't clutter up the `build.sbt`.

File: build.sbt
Comment (armanbilge) [inline build.sbt:83]: I'd like to upstream this in https://github.com/typelevel/sbt-typelevel/issues/239.

File: build.sbt
Comment (armanbilge) [inline build.sbt:84]: This was the most disappointing thing. Seems that Cats has got lots and lots of unused warnings. So we have to disable them if we want fatal warnings for everything else. It would truly be a great thing if someone could chase down and at least `@nowarn` them if they are legit or necessary for bincompat. Then at least we can avoid adding more unused stuff.

File: build.sbt
Comment (armanbilge) [inline build.sbt:326]: This introduces a new API-docs only `cats-docs` module that lets us publish unidocs to Sonatype/Maven. Then they can be browsed at javadoc.io.

Example:
https://www.javadoc.io/doc/org.http4s/http4s-docs_2.13/0.23.11/org/http4s/index.html

File: mima.sbt
Comment (armanbilge) [inline mima.sbt:170]: @dwijnand has very kindly fixed this in https://github.com/lightbend/mima/pull/683 :)

File: build.sbt
Comment (armanbilge) [inline build.sbt:121]: Doctests are supported on Scala 3, but our examples simply don't cross-compile e.g. kind projector syntax for `FunctionK`.

File: build.sbt
Comment (satorg) [reply build.sbt:86]: Seems that Scala 3.0.2 ignores any `@nowarn` annotations, unfortunately. I found it out while working on #4083.

**Comment by armanbilge** [comment]
Ok, looks like we're in business :)

File: build.sbt
Comment (armanbilge) [reply build.sbt:86]: Yes, that's right. But`@nowarn` is supported in Scala 3.1.0 and I believe we should bump for that and other reasons. See:
- https://github.com/typelevel/scalacheck/issues/878

Comment (satorg) [reply build.sbt:86]: We won't be getting any "unused" warnings anymore after this change, is that correct?

Comment (armanbilge) [reply build.sbt:86]: That's right, no more unused warnings :(

Comment (satorg) [reply build.sbt:86]: Hmm... Wouldn't it better to disable the `-Werror` option instead (along with `-Xfatal-?whatever?` used in Scala 2.12)? In this case we will still be seeing warning and hopefully one day will have them fixed :)
Wdyt?

Comment (armanbilge) [reply build.sbt:86]: It's a good question. So IIUC, you mean enable all warnings, but disable fatal warnings?

The only problem is that there are so many unused warnings they drown out everything else maybe.

Comment (satorg) [reply build.sbt:86]: Yes, I also hate seeing warning:) This is why I attempted to make some relief in #4083. And I would be glad to continue working on that.

Answering your question – yes, I think seeing warnings is still better than suppressing them completely.

Comment (armanbilge) [reply build.sbt:86]: Yes, now that I came face-to-face with the problem myself, I truly appreciate your heroic efforts in https://github.com/typelevel/cats/pull/4083 :)

Ok, that's fair. I'll restore all the default compiler flags from sbt-typelevel, but disable fatal warnings.

Comment (satorg) [reply build.sbt:86]: Btw: there're also loads of deprecation warnings in some modules (which mostly don't make any real sense). And there're so many of them that it is quite easy to miss something really important.

Comment (armanbilge) [reply build.sbt:86]: Right, that's exactly my concern. The only way to be sure you don't miss warnings, is to make them fatal. But that means we have to disable the warnings that we can't solve today.

File: mima.sbt
Comment (dwijnand) [reply mima.sbt:170]: Sorry I've not cut a release yet - life things shuffled my priorities the last 3 weeks.

Comment (armanbilge) [reply mima.sbt:170]: No worries, just means we'll appreciate it even more when it arrives! :)

**Comment by armanbilge** [comment]
Ok, for ease of review I put up a PR against this branch that sets up the new site in https://github.com/armanbilge/cats/pull/2. Not much interesting there: adds ~10 lines of build config and then reshuffles files and tweaks links to make Laika happy.

You can preview the new site at: https://armanbilge.github.io/cats

You can preview locally with `sbt docs/tlSitePreview` which starts a server on port 4242. No ruby/jekyll required.

Of course, there is plenty to bikeshed and improve. For example I didn't try very hard to incorporate the cats logos/art in the new site. All ideas welcome, but let's save them for follow up work please? 🙏 😅 The goal is to make it easy to develop the site without having to install external stuff, and easy to deploy it in CI so our changes actually go live :)

**Comment by armanbilge** [comment]
Requesting review from a few folks, whenever you have a chance. Probably the most important decision is what to do about the warnings, see https://github.com/typelevel/cats/pull/4160#discussion_r843162789. Thanks!

**Comment by satorg** [comment]
> Probably the most important decision is what to do about the warnings

Well, from my prospective it is better to keep warnings running. But it is IMO of course. I believe it is a safer way to get rid of them eventually. I already started addressing warnings and managed to complete quite a chunk of job in #4083. And I'd be glad to keep it up. So once we'll have wiped the remaining warnings out we can safely turn the "fatal" option back.

**Comment by armanbilge** [comment]
I don't want to bring things to standstill, but FYI merging other PRs is a bit hard on this one :)

File: project/plugins.sbt
Comment (rossabaker) [reply project/plugins.sbt:4]: I don't see a particular reason not to bite the bullet and add the headers, though it would be nice to get this substantial work to a landing.

Comment (armanbilge) [reply project/plugins.sbt:4]: Yep, super easy to make that change. Mostly it just creates a massive diff :)

Also, I wasn't completely certain what to make of Cats' [COPYING](https://github.com/typelevel/cats/blob/main/COPYING) which includes a Scalaz notice as well. E.g. whether it's appropriate to put a blanket license over every file.

Comment (rossabaker) [reply project/plugins.sbt:4]: In http4s, we generate our copyright and also preserve the original copyright on derived files.  This is a fairly common practice, and one that doesn't irritate the tooling.

Comment (armanbilge) [reply project/plugins.sbt:4]: Right, http4s has been very diligent about that which helped. But AFAICT not a single file in Cats has a copyright header, whether its derived or not.

Comment (rossabaker) [reply project/plugins.sbt:4]: Oh, I see.  I don't know how we'd go back and add that now.  I say we made a good faith effort with the blanket COPYING file, and I'd be happy to add a second header if anyone cares enough to point out the derived works.

**Review by rossabaker — DISMISSED** [review]
This is wonderful.

**Review by rossabaker — APPROVED** [review]
I hope I didn't piss off everyone welcoming headers, but they're almost everywhere else in Typelevel...

**Comment by armanbilge** [comment]
Yep, rolled out 16k lines of headers 😬 Also applied Sergei's suggestion to keep all the warnings enabled, non-fatally. Merged in my separate site PR.

Oh, I need to set a proper start year for the headers ...

**Comment by satorg** [comment]
> rolled out 16k lines of headers
915 files changed versus ~140 before that.. Seems that Git has lost tracks of small files after they received the headers..

What are the key files for the review? I doubt if I can make all 915 :)

**Comment by armanbilge** [comment]
`build.sbt` is really the key file. But honestly sbt is hard to understand for a good review :) best way to review IMO is to interact with the sbt console. Check that settings are configured the way you expect. Check that MiMa previous versions includes what it should (make sure tags on your local are updated). Try previewing the website with `docs/tlSitePreview`. Stuff like that :)

**Comment by armanbilge** [comment]
Any thoughts on landing this one? 🙏 I need to update this PR to account for the just-merged https://github.com/typelevel/cats/pull/4166 ...

**Comment by satorg** [comment]
> Any thoughts on landing this one? 

Agree. I would vote to hold off other merges until we get this landed.

Regarding the merging of this one.... I found it tremendously difficult to reason about the safety of the changes made. They look good to me, but I realize that I cannot follow everything here.

So... What about scheduling an "RC" release after merging this PR? E.g. "2.8.0-RC1" or something. Just to let everyone (who would like to participate) test the result artifacts for potential BinCompat issues first. Would it make sense?

**Comment by armanbilge** [comment]
Sure, I'm 👍 to publishing Ms, RCs, etc. :) actually, as soon as this merges it should immediately publish a snapshot if all goes well.

> Just to let everyone (who would like to participate) test the result artifacts for potential BinCompat issues first. Would it make sense?

Depends on what you want to test :) if you want to test the fix for https://github.com/typelevel/cats/issues/4062, for sure, that makes sense.

If you want to test whether the build will prevent the publishing of binary-breaking changes, then we'd have to try and introduce one first :) because besides the fix for https://github.com/typelevel/cats/issues/4062 this doesn't really change any code.

**Review by satorg — APPROVED** [review]
I think we should go ahead and merge it.

**Comment by armanbilge** [comment]
https://s01.oss.sonatype.org/content/repositories/snapshots/org/typelevel/cats-core_3/

**Comment by satorg** [comment]
@armanbilge now when I run `sbt +prePR` on main I receive a lot of errors like these below:
```
[error] /Users/satorg/projects/typelevel/cats/algebra-laws/shared/src/test/scala/algebra/laws/LawTests.scala:74:23: object DeMorganLaws in package laws is deprecated (since 2.7.0): Laws moved to LogicLaws
[error]   checkAll("Boolean", DeMorganLaws[Boolean].deMorgan(DeMorgan.fromBool(Bool[Boolean])))
[error]                       ^
[error] /Users/satorg/projects/typelevel/cats/algebra-laws/shared/src/test/scala/algebra/laws/LawTests.scala:189:26: object GroupLaws in package laws is deprecated (since 2.7.0): Provided by cats.kernel.laws
[error]     checkAll("Order[N]", GroupLaws[Order[N]].monoid)
[error]                          ^
[error] /Users/satorg/projects/typelevel/cats/algebra-laws/shared/src/test/scala/algebra/laws/LawTests.scala:193:25: object GroupLaws in package laws is deprecated (since 2.7.0): Provided by cats.kernel.laws
[error]       checkAll("Eq[N]", GroupLaws[Eq[N]].boundedSemilattice)
[error]                         ^
[error] /Users/satorg/projects/typelevel/cats/algebra-laws/shared/src/test/scala/algebra/laws/LawTests.scala:197:25: object GroupLaws in package laws is deprecated (since 2.7.0): Provided by cats.kernel.laws
[error]       checkAll("Eq[N]", GroupLaws[Eq[N]].semilattice)
[error]                         ^
```
Just to clarify: is it supposed to run this way?

**Comment by armanbilge** [comment]
@satorg hmm, no, that's a bug 😕 try setting `ThisBuild / tlFatalWarningsInCi := false`.

**Comment by satorg** [comment]
Unfortunately, changing values of either `tlFatalWarnings` or `tlFatalWarningsInCi` does not work since `prePR` reloads the project before starting doing everything else....

**Comment by armanbilge** [comment]
Sorry, I meant we should permanently add it to `build.sbt`. It's my fault for not setting this up right to begin with :)

**Comment by satorg** [comment]
No worries. I can go ahead and file a PR for that.

**Comment by satorg** [comment]
@armanbilge although it is merged now... But seems I missed it while was reviewing.
https://github.com/typelevel/cats/blob/main/build.sbt#L26

It sets the default Scala version to 2.12, which is different from how it was before (2.13).
I wonder was it made for some specific reason? Because to me it seems less convenient since requires switching versions manually more often while compiling from SBT command line. Moreover, Metals will use 2.12 by default to generate Bloop files.

**Comment by armanbilge** [comment]
@satorg I agree, 2.13 makes more sense. I changed it to 2.12 in ce873ff9f74d78b4eab975e78e65b06c0d580fbd because mdoc wasn't working in 2.13, since the sbt-typelevel-site plugin uses the default Scala for building/deploying documentation. In principle, we can hack around this in the build ... but just like fatal warnings, best option is to fix the root problem. I.e. get the docs building on Scala 2.13. We can make an issue for it and a contributor may be interested to work on it.

**Comment by satorg** [comment]
Ah I see – it was for the reason.
Yeah it is definitely worth the fixing. I'm not sure if it is going to be a "good first issue" though... :)

**Pull request merged**
